honeybee/model.py

```diff
@@ -119,243 +207,391 @@ def north_vector(self, value):
+                object will be rotated.
+        """
+        for room in self._rooms:
+            room.rotate(axis, angle, origin)
+        for face in self._faces:
+            face.rotate(axis, angle, origin)
+        for shade in self._shades:
+            shade.rotate(axis, angle, origin)
+        for aperture in self._apertures:
+            aperture.rotate(axis, angle, origin)
+        for door in self._doors:
+            door.rotate(axis, angle, origin)
+
+    def rotate_xy(self, angle, origin):
+        """Rotate this Model counterclockwise in the world XY plane by a certain angle.
+
+        Args:
+            angle: An angle in degrees.
+            origin: A ladybug_geometry Point3D for the origin around which the
+                object will be rotated.
+        """
+        for room in self._rooms:
+            room.rotate_xy(angle, origin)
+        for face in self._faces:
+            face.rotate_xy(angle, origin)
+        for shade in self._shades:
+            shade.rotate_xy(angle, origin)
+        for aperture in self._apertures:
+            aperture.rotate_xy(angle, origin)
+        for door in self._doors:
+            door.rotate_xy(angle, origin)
+
+    def reflect(self, plane):
+        """Reflect this Model across a plane with the input normal vector and origin.
+
+        Args:
+            plane: A ladybug_geometry Plane across which the object will
+                be reflected.
+        """
+        for room in self._rooms:
+            room.reflect(plane)
+        for face in self._faces:
+            face.reflect(plane)
+        for shade in self._shades:
+            shade.reflect(plane)
+        for aperture in self._apertures:
+            aperture.reflect(plane)
+        for door in self._doors:
+            door.reflect(plane)
+
+    def scale(self, factor, origin=None):
+        """Scale this Model by a factor from an origin point.
+
+        Args:
+            factor: A number representing how much the object should be scaled.
+            origin: A ladybug_geometry Point3D representing the origin from which
+                to scale. If None, it will be scaled from the World origin (0, 0, 0).
+        """
+        for room in self._rooms:
+            room.scale(factor, origin)
+        for face in self._faces:
+            face.scale(factor, origin)
+        for shade in self._shades:
+            shade.scale(factor, origin)
+        for aperture in self._apertures:
+            aperture.scale(factor, origin)
+        for door in self._doors:
+            door.scale(factor, origin)
 
     def check_duplicate_room_names(self, raise_exception=True):
-        """Check that there are no duplicate room names in the model."""
-        room_count = len(self._room_names)
-        unique_room_count = len(set(self._room_names))
-        if room_count != unique_room_count:
+        """Check that there are no duplicate Room names in the model."""
+        room_names = []
```

Comment:
You probably want to use `set` in this case which will make the look-up much faster. This is also a good to see:

https://stackoverflow.com/questions/25706136/efficiently-find-repeated-characters-in-a-string/25706298#25706298

Reply:
This comment applies to all check for duplicate items.

Reply:
The main reason why it is implemented this way is because the exact next question that you have after you run this check and get a failure is "what are all of the duplicated names in my model?"  This method allows you to get all of them.

If you think it makes more sense, I can separate each of these into two methods: one that checks whether there are any duplicate names and another that returns a list of all the duplicate names.  I guess this would allow the best of both worlds by allowing for fast performance when the model is good but giving you all of the information you need to fix your model when it is not good.

Also, according to that discussion that you linked to, doesn't using `set` give you worse performance? Isn't using sort or dict better?

Reply:
Your code stays almost the same so you can do all you are doing right now.

```python

    def check_duplicate_room_names(self, raise_exception=True):
        """Check that there are no duplicate Room names in the model."""
        room_names = set()   # this line is changes
        duplicate_names = []
        for room in self._rooms:
            if room.name not in room_names:  # now this is much faster than looking up in a list
                room_names.add(room.name)  # this is also changed
            else:
                duplicate_names.append(room.name)
        if duplicate_names != []:
            if raise_exception:
                raise ValueError('The model has the following duplicated '
                                 'Room names:\n{}'.format('\n'.join(duplicate_names)))
            return False
        return True
```

sort works well with strings or numbers not objects and dict might be too much for what you are trying to do here. Set will give you a much better performance than list for look-ups which will be the main bottleneck here.

Reply:
Also you should probably use set(duplicates) for error otherwise you might get several counts of the same name if the name has been used more than two times.

Reply:
Ok, what you have looks good but, to clarify, we are talking about strings here NOT objects (room names are strings).  So sort might still be the faster option but I also like this implementation with sets.

Reply:
Ok, I just implemented it with `set` as you suggested.  Just let me know if you want me to change it for better performance.

Reply:
To be able to use sorted you need to create the list of names first from the Room objects and then use sorted and so on. You can also use sorted with `key` and then compare `.names` but still you are dealing with objects.

Reply:
I see what you mean.  Ok, then it seems like `set()` gives us a smaller memory footprint, then, so let's go with that.